### PR TITLE
New package: lc0

### PR DESCRIPTION
**Pull request by shtayerc**

### CODE REVIEW

**Comment by shtayerc** [comment]
For xlint error there is already [pull request](https://github.com/leahneukirchen/xtools/pull/173). I would need some help with cross compiling.

**Comment by ericonr** [comment]
Why not make the backend choice a build option? Defaulting to openblas seems reasonable too.

File: srcpkgs/lc0/template [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+maintainer="shtayerc <david.murko@mailbox.org>"
+license="GPL-3.0-only"
+homepage="http://lczero.org"
+distfiles="https://github.com/LeelaChessZero/lc0/archive/v${version}.tar.gz
+ https://github.com/LeelaChessZero/lczero-common/archive/${_githash}.tar.gz"
+checksum="2f38cd5fb1822798d10d806fe8930f60a5b3a7f92b9dfcbbeb210df8a8b37b86
+ c3de32a249b25f092218b7550ce00134998cf3fef8ebbb08801883ed48f36576"
+
+build_options="openblas opencl"
+build_options_default="openblas"
```
Comment (ericonr) [inline srcpkgs/lc0/template:23]: Assuming you can't have both backends at the same time, `vopt_conflict openblas opencl`

File: srcpkgs/lc0/template [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Template file for 'lc0'
+pkgname=lc0
+version=0.27.0
+revision=1
+wrksrc="lc0-${version}"
+build_style=meson
+configure_args="-Dgtest=false --buildtype=release"
+meson_builddir="build/release"
```
Comment (ericonr) [inline srcpkgs/lc0/template:8]: Why force release?

What's special about `build/release`?

File: srcpkgs/lc0/template [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+_githash=00fd892e648160c294346c87449126d9bad80a16
+hostmakedepends="tar pkg-config"
+makedepends="$(vopt_if openblas openblas-devel)
+ $(vopt_if opencl ocl-icd-devel) protobuf-devel zlib-devel eigen"
```
Comment (ericonr) [inline srcpkgs/lc0/template:12]: Style nit: it's cleaner to have the build options after the normal deps.

File: srcpkgs/lc0/template [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+if [ "$CROSS_BUILD" ] && [ "$build_option_openblas" ]; then
+	configure_args+=" -Dopenblas_include=${XBPS_CROSS_BASE}/usr/include/openblas/"
+fi
+
+if [ "$XBPS_TARGET_NO_ATOMIC8" ]; then
+	makedepends+=" libatomic-devel"
+fi
+
+post_extract() {
+	cp -a ../lczero-common-${_githash}/proto/ ./libs/lczero-common/
```
Comment (ericonr) [inline srcpkgs/lc0/template:34]: Please use some form of `mv` so the directory isn't left after the build is done. something like `rmdir libs/lczero-common` then `mv ../lczero-common-${_githash} libs/lczero-common` should work.

Also, consider being specific: `_lczero_commit`

File: srcpkgs/lc0/template [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+do_install() {
+	vbin build/release/lc0 lc0
```
Comment (ericonr) [inline srcpkgs/lc0/template:38]: Does it lack a install target?

File: srcpkgs/lc0/template [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# Template file for 'lc0'
+pkgname=lc0
+version=0.27.0
+revision=1
+wrksrc="lc0-${version}"
```
Comment (ericonr) [inline srcpkgs/lc0/template:5]: This is now the default `wrksrc`, and can be ommitted ;)

File: srcpkgs/lc0/template (hunk 2 above)
Comment (shtayerc) [reply srcpkgs/lc0/template]: I guess nothing, it works fine without specifying it.

File: srcpkgs/lc0/template [hunk 7]
```diff
@@ -0,0 +1,35 @@
+# Template file for 'lc0'
+pkgname=lc0
+version=0.27.0
+revision=1
+build_style=meson
+configure_args="-Dgtest=false"
+_lczero_commit=00fd892e648160c294346c87449126d9bad80a16
+hostmakedepends="tar pkg-config"
```
Comment (ericonr) [inline srcpkgs/lc0/template:8]: Last nit, does it actually need `tar` here? It shouldn't be using any subproject, after all.

Comment (shtayerc) [reply srcpkgs/lc0/template]: Done

**Comment by ericonr** [comment]
So, what GUI are you using this with? I just tried using with `cutechess` and it just hangs...

https://lczero.org/play/quickstart/ mentions a neural network file, aren't we missing that?

**Comment by ericonr** [comment]
```
open("/home/ericonr/.local/share/lc0", O_RDONLY|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
open("/home/ericonr/.local/share/lc0/networks", O_RDONLY|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
open("/usr/local/share/lc0", O_RDONLY|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
open("/usr/local/share/lc0/networks", O_RDONLY|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
open("/usr/share/lc0", O_RDONLY|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
open("/usr/share/lc0/networks", O_RDONLY|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
```

Could we install a default network in `/usr/share/lc0/networks`? Is it easy to download one as a distfile?

**Comment by ericonr** [comment]
Maybe https://training.lczero.org/get_network?sha=b30e742bcfd905815e0e7dbd4e1bafb41ade748f85d006b8e28758f1a3107ae3
for 703810.pb.gz

**Comment by ericonr** [comment]
Or should users be expected to know how to install things?

**Comment by shtayerc** [comment]
I am using lc0 with my own gui https://github.com/shtayerc/cboard, I do not know how to configure it for qutechess. But you can test it without gui, but you need weights file. You can find it here https://github.com/LeelaChessZero/lc0/wiki/Best-Nets-for-Lc0 . I think users should decide for themself, because nets are meant for different things and there isn't one good for all.
Then you run lc0 from command line like that `lc0 --weights=/path/to/file`
And after you type `go infinite` it should be calculating.

File: srcpkgs/lc0/template [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+build_style=meson
+configure_args="-Dgtest=false"
+_lczero_commit=5680c5fad9f3b52288d67f738b272fd09de5ee0b
+hostmakedepends="pkg-config"
+makedepends="protobuf-devel zlib-devel eigen $(vopt_if openblas openblas-devel)
+ $(vopt_if opencl ocl-icd-devel)"
+short_desc="UCI chess engine designed to play via neural network"
+maintainer="shtayerc <david.murko@mailbox.org>"
+license="GPL-3.0-only"
+homepage="http://lczero.org"
```
Comment (ericonr) [inline srcpkgs/lc0/template:14]: Use https

**Review by ericonr — COMMENTED** [review]
Should be good to merge once homepage is fixed.

File: srcpkgs/lc0/template (hunk 8 above)
Comment (shtayerc) [reply srcpkgs/lc0/template]: @ericonr done

**Pull request merged**